PermissionsScraper/PermissionsScraper/Services/PermissionsProcessor.cs

```diff
@@ -0,0 +1,149 @@
+        /// <param name="permissionsDescriptionsText">The string input with permissions descriptions.</param>
+        /// <param name="referencePermissionsDictionary">The target permissions descriptions dictionary which the extracted permissions will be added into.</param>
+        public static void ExtractPermissionsDescriptionsIntoDictionary(string[] scopesNames,
+                                                                        string permissionsDescriptionsText,
+                                                                        ref Dictionary<string, List<Dictionary<string, object>>> referencePermissionsDictionary,
+                                                                        string topLevelDictionaryName = null)
+        {
+            UtilityFunctions.CheckArgumentNull(scopesNames, nameof(scopesNames));
+            UtilityFunctions.CheckArgumentNull(referencePermissionsDictionary, nameof(referencePermissionsDictionary));
+            UtilityFunctions.CheckArgumentNullOrEmpty(permissionsDescriptionsText, nameof(permissionsDescriptionsText));
+
+            var permissionsDescriptionsToken = topLevelDictionaryName is null
+                ? JsonConvert.DeserializeObject<JObject>(permissionsDescriptionsText)
+                : JsonConvert.DeserializeObject<JObject>(permissionsDescriptionsText).Value<JArray>(topLevelDictionaryName)?.First;
+
+            ExtractPermissionsDescriptionsIntoDictionary(scopesNames, permissionsDescriptionsToken, ref referencePermissionsDictionary);
+        }
+
+        /// <summary>
+        /// Extracts permissions descriptions from a <see cref="JToken"/> source
+        /// and adds them to a target permissions descriptions dictionary.
+        /// </summary>
+        /// <param name="scopesNames">The scope names to retrieve from the permissions descriptions.</param>
+        /// <param name="permissionsDescriptionsToken">The <see cref="JToken"/> input with permissions descriptions.</param>
+        /// <param name="referencePermissionsDescriptions">The target permissions descriptions dictionary which the extracted permissions will be added into.</param>
+        private static void ExtractPermissionsDescriptionsIntoDictionary(string[] scopesNames,
+                                                                         JToken permissionsDescriptionsToken,
+                                                                         ref Dictionary<string, List<Dictionary<string, object>>> referencePermissionsDescriptions)
+        {
+            UtilityFunctions.CheckArgumentNull(scopesNames, nameof(scopesNames));
+            UtilityFunctions.CheckArgumentNull(permissionsDescriptionsToken, nameof(permissionsDescriptionsToken));
+            UtilityFunctions.CheckArgumentNull(referencePermissionsDescriptions, nameof(referencePermissionsDescriptions));
+
+            foreach (string scopeName in scopesNames)
```

Comment:
all of that could be replaced by a few linq statements...

Reply:
In this PR [comment](https://github.com/microsoftgraph/microsoft-graph-devx-api/pull/636#discussion_r669058046) you did mention that semantically, `foreach` would signify that we want to perform an _action_ for each of the element in the `IEnumerable`, whereas a `linq` `Select` signifies we're _converting_ each element of an the `IEnumerable` to a different thing. In our case we intend to perform several _actions_ using our element.

Reply:
from reading the code it looks to me that these loops are projecting two collections, retaining elements of the projection based on conditions. In my experience linq is really good at dataset manipulation (projections, correlations, filtering,...) and is easier to read than loops for that (besides the potential performance gains). I do reserve loops for actions, and by actions I mean calling methods on subobjects/classes.

Reply:
I've tried replacing the `foreach` with `Linq`, however the `ref` params trigger the error message: "_Cannot use ref, out, or in parameter inside an anonymous method, lambda expression, query expression or local function._"